File: q1458/q1458-top-down.java
class Solution {
    public int maxDotProduct(int[] nums1, int[] nums2) {
        int[][] memo = new int[nums1.length][nums2.length];
        for (int[] row : memo) {
            Arrays.fill(row, Integer.MIN_VALUE);
        }
        return dpHelper(nums1, nums2, 0, 0, memo);
    }

    private int dpHelper(final int[] nums1, final int[] nums2, int i, int j, int[][] memo) {
        // Base case: Check if we are at the end of either nums1 or nums2
        if (i >= nums1.length || j >= nums2.length) {
            return Integer.MIN_VALUE; // Return a very small value as invalid
        }

        // Memoized case
        if (memo[i][j] != Integer.MIN_VALUE) {
            return memo[i][j];
        }

        // Calculate the possible solutions
        int sol1 = nums1[i] * nums2[j] + Math.max(0, dpHelper(nums1, nums2, i + 1, j + 1, memo));
        int sol2 = dpHelper(nums1, nums2, i + 1, j, memo);
        int sol3 = dpHelper(nums1, nums2, i, j + 1, memo);
        memo[i][j] = Math.max(sol1, Math.max(sol2, sol3));
        return memo[i][j];
    }
}

File: q1458/q1458-top-down.cpp
class Solution
{
public:
    int maxDotProduct(vector<int> &nums1, vector<int> &nums2)
    {
        vector<vector<int>> memo(nums1.size(), vector<int>(nums2.size(), INT_MIN));
        return dpHelper(nums1, nums2, nums1.size() - 1, nums2.size() - 1, memo);
    }

private:
    int dpHelper(const vector<int> &nums1, const vector<int> &nums2, const int i, const int j, vector<vector<int>> &memo)
    {

        if (i < 0 || j < 0)
            return INT_MIN;

        if (memo[i][j] != INT_MIN)
            return memo[i][j];

        int sol1 = max(0, dpHelper(nums1, nums2, i - 1, j - 1, memo)) + nums1[i] * nums2[j];
        int sol2 = dpHelper(nums1, nums2, i - 1, j, memo);
        int sol3 = dpHelper(nums1, nums2, i, j - 1, memo);
        int res = max(sol1, max(sol2, sol3));
        memo[i][j] = res;
        return res;
    }
};